Add a sphinx docs structure for hosting gmt examples

meghanrjones opened the pull request:
This PR sets up a sphinx project for hosting the GMT simple examples and tutorials.

Relates to https://github.com/GenericMappingTools/gmt/pull/5364.

meghanrjones commented:
@seisman, is 'Import Git Repository' in Vercel the proper way to setup the vercel previews for this?

seisman commented:
> @seisman, is 'Import Git Repository' in Vercel the proper way to setup the vercel previews for this?

I think yes.

meghanrjones commented:
This PR adds a sphinx documentation setup to host simple gallery examples and tutorials in a separate repository from gmt, which will hopefully lead to more new contributors.

Here is a preview for the built documentation: https://gmt-examples-j26ucrg4b-gmt.vercel.app/.

The sections can be changed as needed; I went based on the [PyGMT gallery](https://www.pygmt.org/dev/gallery/index.html) as a first draft. I also translated one gallery example and one tutorial from PyGMT to demonstrate the structure.

Here's a preview of the instructions for adding new examples: https://gmt-examples-j26ucrg4b-gmt.vercel.app/contributing.html#contributing-new-examples.

Let me know what you think @seisman @PaulWessel @joa-quim @Esteban82 @gd-a @anbj and anyone else who wants to provide feedback 🙂

joa-quim commented:
Why are you recommending the usage of Conda and Conda environments? Conda is a overbloated s that eats Gigs and Gigs of user disk space. Last time I tried an env it created a > 3.5 GB copy just for run a small py package. I have no intention to use Conda other than on a experimenting ground and at least users should be aware of this.

PaulWessel commented:
I think the setup needs to allow for old folk like me who do not want to use Conda as well, for different reasons than @joa-quim.  So probably an either or, or a list of ways.

anbj commented:
I think it's a good start.
I haven't really tried conda, but I'm confused why this is needed.
Why would you need an isolated environment for this?
Just concerned about not *keeping it simple, stupid*.

meghanrjones commented:
I added instructions for installing dependencies (apart from GMT) using pip and took away the 'recommendation'. If you use pip, you can follow those instructions and do not need to start using conda. Relying on virtual environments to isolate versions is a preference, not required.

joa-quim commented:
> for different reasons than @joa-quim.

It's not **my** reasons. I have a 1TB disk and can afford these craziness but I see it around me all the time. People with laptops work directly on external disks because they have no space left. And then come the  _Condas_ of this world who think spending 3 GB for something that should take 30 MB (or 30 even kB) is cool.

PaulWessel commented:
> > for different reasons than @joa-quim.
> 
> It's not **my** reasons. I have a 1TB disk and can afford these craziness but I see it around me all the time. People with laptops work directly on external disks because they have no space left. And then come the _Condas_ of this world who think spending 3 GB for something that should take 30 MB (or even 30 kB) is cool.

Fair enough. Though I speak as the owner of a 14" laptop with 8 Tb of SSD space.

joa-quim commented:
> with 8 Tb of SSD space.

Yep, but half oh it is booked for swapping all those memory leaks of the fancy M's 😄

joa-quim commented:
> I added instructions for installing dependencies (apart from GMT) using pip and took away the 'recommendation'. 

But why is Python even involved here? I don't agree it that at all. Be it Conda or whatever.
Contributors for sure have GMT installed already and know how to use it so o need for any recommendations on that side.

meghanrjones commented:
> > I added instructions for installing dependencies (apart from GMT) using pip and took away the 'recommendation'.
> 
> But why is Python even involved here? I don't agree it that at all. Be it Conda or whatever. Contributors for sure have GMT installed already and know how to use it so no need for any recommendations on that side.

What documentation builder do you want us to use here instead? sphinx is used already for the gmt docs and website, so i do not understand why it would be fine for those but not this.

PaulWessel commented:
I think @joa-quim's point is that one does not need conda to run sphinx.

meghanrjones commented:
Yes, conda is not required and python is required. The 'official' installation instruction for sphinx uses pip: https://github.com/sphinx-doc/sphinx#installation

joa-quim commented:
OK, I must red more before speak but what I see in the preview is still

```We highly recommend using Anaconda and the conda package manager to install and manage your packages.```

So I thought, even after changing to ``pip`` that the recommendations were about the GMT installation.
However, from what I understand now, contributors are required to have a full working sphinx installation to be able to contribute an example. I think that is asking too much to people. They can do that if they want but they should also be allowed to contribute only the script an, if required, the companion data.

meghanrjones commented:
Here's the latest preview for that section: https://gmt-examples-b5g9rmfb8-gmt.vercel.app/contributing.html.

I'll look into making it simpler for people who don't want to spend time with sphinx, ReST, etc.

Regarding the companion data, do you think those should get hosted along with this repository or the GMT server's cache directory? I had envisioned the examples using existing remote files, since opening the possibility for people to add new datasets for the examples could easily get out of hand in terms of size and maintenance efforts IMO.

joa-quim commented:
Yes, companion data can be a problem. Maybe we can accept new data if it's only in tabular form and impose a limit. Regarding grids and images we should start by saying that people must use the existing remote files, or other files accessible with GDAL's ``/vsicurl/`` mechanism.

PaulWessel commented:
> Yes, companion data can be a problem. Maybe we can accept new data if it's only in tabular form and impose a limit. Regarding grids and images we should start by saying that people must use the existing remote files, or other files accessible with GDAL's `/vsicurl/` mechanism.

Or just plain URL which GMT can read.

seisman commented on docs/contributing.md:
````diff
@@ -0,0 +1,143 @@
+conda env create
+```
+
+Before building and testing the project, you have to activate the environment
````
```suggestion
Before building the documentation, you have to activate the environment
```

seisman commented on docs/contributing.md:
```diff
@@ -0,0 +1,143 @@
+
+There are two main components to GMT examples project:
+
+* Gallery examples, with source material in the `docs/examples/gallery/` folder.
+* Tutorial examples, with source material in the `docs/examples/tutorials/` folder.
```
Why not putting gallery and tutorial examples directly in the `docs` folder, i.e., `docs/gallery` and `docs/tutorials`.

seisman commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+myst-parser
```
The myst-parser extension is used to support markdown source files. Currently, the only markdown file is `docs/contributing.md`. We can remove the myst-parser dependency if we convert contributing.md from markdown to rst.

seisman commented on docs/contributing.md:
````diff
@@ -0,0 +1,143 @@
+
+  ```bash
+  .. gmtplot::
+     :language: bash
````
I think most examples (if not all) will in written in Bash. So, it makes sense to set the default language to "bash", so contributors don't have to specify the language option explicitly. The default language is set by the `highlight_language` setting in the conf.py, as mentioned in https://www.generic-mapping-tools.org/sphinx_gmt/latest/api/generated/sphinx_gmt.gmtplot.html#module-sphinx_gmt.gmtplot.

seisman commented on docs/contributing.md:
````diff
@@ -0,0 +1,143 @@
+  The figures will be placed after the source code in the built documentation
+  by the `sphinx_gmt` extension.
+
+* Add the file to the appropriate section in `docs/index.rst` using the following
+  template:
+
+  ```bash
+    -  .. image:: _images/<file-name>-gmtplot-0.png
+        :target: examples/gallery/<section>/<file-name>.html
+        :width: 80%
+        :align: center
+
+     :doc:`examples/gallery/<section>/<file-name>`
+  ```
````
I think we need a better way to display the images on the landing page, similar to the Sphinx-Gallery extension, but it definitely needs more work, so it can be done in other RPs.

meghanrjones replied:
Do you mean that the appearance should be made better or that we should figure out a way to automatically populate the index page based on the files in `docs/gallery` and `docs/tutorials`?

meghanrjones commented on docs/conf.py:
```diff
@@ -0,0 +1,64 @@
+# -- Project information -----------------------------------------------------
+
+project = 'GMT Examples'
+copyright = '2022, The GMT Team'
+author = 'The GMT Team'
```
Thanks for all the comments!

What about having the copyright as "The GMT Examples Authors" and adding an authorship policy similar to PyGMT, whereby anyone who contributes can add themselves to an AUTHORS.md file?

seisman replied:
> What about having the copyright as "The GMT Examples Authors" and adding an authorship policy similar to PyGMT, whereby anyone who contributes can add themselves to an AUTHORS.md file?

Sounds good to me.

leouieda commented:
👏🏽 awesome work on this @meghanrjones! Makes a lot of sense to have a separate set of user contributed examples. And it's a good way to get them used to the setup and being able to contribute to the docs afterwards.

Just FYI I looked into what it would take to get sphinx-gallery to work with a bunch of bash scripts instead of Python scripts a while ago. Capturing the images is easy enough but getting it run the code and create the rst files is trickier. There was a discussion of this in sphinx-gallery/sphinx-gallery#421 (using the Jupyter machinery to execute code in basically any language) but doesn't seem to have been implemented yet.

leouieda commented on docs/contributing.md:
```diff
@@ -0,0 +1,143 @@
+- [sphinx_gmt](https://www.generic-mapping-tools.org/sphinx_gmt/latest/)
+- [sphinx_rtd_theme](https://sphinx-rtd-theme.readthedocs.io/en/stable/installing.html)
+
+Two options for installing the dependencies are [pip](#pip-setup) and [conda](#conda-setup).
```
Might be worth mentioning here that sphinx and the other things needed to build the website are Python based and this is why Python is needed (and pip and conda are options).

leouieda replied:
Also mentioning that they'll need a working Python environment, however they want to set one up. Then we can say that if they don't know how to do that, installing [Miniforge](https://github.com/conda-forge/miniforge) and using the `conda` environment below is the easiest way to get everything setup. If they know what they're doing, then they can do whatever and avoid conda if desired.

I recommend miniforge here instead of Anaconda because it really only comes with the bare minimum and is (somewhat) independent from Anaconda Inc.

leouieda commented:
Since there is nothing here besides the sphinx website sources, it may be worth putting everything in the top-level directory instead of `docs`, like we do on https://github.com/GenericMappingTools/website/. Entirely optional, though.

leouieda commented:
> Or just plain URL which GMT can read.

This can be tricky depending on the URL. If the data moves, these docs will break and it will be out of our control. 

Alternatively, getting their data and uploading somewhere we control (server, Zenodo, whatever) runs into the problem of open licenses (and mostly the lack thereof).

joa-quim commented:
> I recommend miniforge here instead of Anaconda because it really only comes with the bare minimum and is (somewhat) independent from Anaconda Inc.

Hi Leonardo,
I am sorry but I never saw that (and reinstalled my _miniconda_ (that was using 18 GB) not long ago. It may be true on first install (if user refuses to install what is proposed, don't remember) but it quickly jumps to the multi Gigs. Mine has already *recovered* to 8 GB.

joa-quim commented:
Another thing. It would be nice to anticipate that the examples can be shown both in pure GMT as well as GMT.jl, PyGMT and eventually GMTMEX. If we don't have many it won't be much work to translate between the syntaxes and PyGMT already has quite a few examples that can be used here.

meghanrjones commented:
One way to support both pure GMT and the wrappers would be to have a toggle button dropdown that contains the equivalent source code in GMT.jl, PyGMT, GMT/MEX. This would be easy to implement. It would be more difficult to host entire separate tutorials/gallery examples for wrappers, mostly because GMT and the wrappers would all need to be installed for the continuous documentation which can be fragile/slow based on [try-gmt](https://github.com/GenericMappingTools/try-gmt).

PaulWessel commented:
Would we need to support the running of all those wrapper scripts in the CI or do we just support the bash script and simply display the comparable wrapper scripts as is?

joa-quim commented:
I was thinking in just providing the the equivalent scripts. Not the burden of setting up an infrastructure to make them pass CI tests. At least not at the beginning. A dropdown menu like the one for selecting different versions of the manual would be good.

meghanrjones commented:
That sounds good, I can add instructions in a separate PR.

seisman commented on docs/contributing.rst:
```diff
@@ -0,0 +1,143 @@
+Unix/macOS::
+
+  python -m pip install -r requirements.txt
+
```
```suggestion
```

seisman commented on docs/contributing.rst:
```diff
@@ -0,0 +1,143 @@
+environment from the ``environment.yml`` file::
+
+  conda env create
+
```
```suggestion
```

seisman commented on docs/contributing.rst:
```diff
@@ -0,0 +1,143 @@
+* Create a new ``.rst`` file inside the appropriate sub-directory in
+  ``docs/gallery/`` or ``docs/tutorials/``.
+* Add a descriptive title and as much explanation as necessary.
+* Add hyperlinks to GMT modules using the ``gmt-module`` restructured text directive::
```
```suggestion
* Add hyperlinks to GMT modules using the ``gmt-module`` reStructuredText directive::
```

seisman commented on docs/contributing.rst:
```diff
@@ -0,0 +1,143 @@
+
+  ::
+
+    .. gmtplot::
+
+      gmt begin basemap png
+        gmt basemap -B -Rg -JH5c
+      gmt end show
```
reStructureText needs 3 or 4 leading whitespaces.
```suggestion
    .. gmtplot::

       gmt begin basemap png
         gmt basemap -B -Rg -JH5c
       gmt end show
```

seisman commented on docs/contributing.rst:
````diff
@@ -0,0 +1,143 @@
+
+  ::
+
+    ```bash
+      -  .. image:: _images/<file-name>-gmtplot-0.png
+          :target: gallery/<section>/<file-name>.html
+          :width: 80%
+          :align: center
+
+       :doc:`gallery/<section>/<file-name>`
+    ```
````
```suggestion
    ```bash
      -  .. image:: _images/<file-name>-gmtplot-0.png
            :target: gallery/<section>/<file-name>.html
            :width: 80%
            :align: center

         :doc:`gallery/<section>/<file-name>`
    ```
```

meghanrjones commented:
I've addressed all the issues raised in the PR comments except these four:

> Allow users to contribute scripts without needing a working sphinx installation

I'll work on instructions for how to contribute an example without a local development environment in a separate PR.

> Optionally put everything in a top-level directory

I prefer to keep the top-level directory relatively small.

> Establish guidelines for datasets used in gmt examples

I'll submit a separate PR with a new contributing guide section for further discussion.

> Add toggle buttons to support GMT.jl, PyGMT, and GMTMEX equivalent scripts

I'll create an issue to track this, which can be added in a separate PR.

In the meantime, I'll merge this PR so we have a starting point and will transfer any issues from the GMT repository that can be more appropriately addressed through the GMT examples pages.
